base/range.jl

```diff
@@ -1486,3 +1486,131 @@ julia> mod(3, 0:2)  # mod(3, 3)
+between the given endpoints. That is, the ratio of successive elements is
+a constant, calculated from the length.
+
+Like [`LinRange`](@ref), the first and last elements will be exactly those
+provided, but intermediate values may have small floating-point errors
+(especially for complex numbers).
+
+This is similar to `geomspace` in Python. Unlike `PowerRange` in Mathematica,
+you specify the number of elements not the ratio.
+Unlike `logspace` in Python and Matlab, the `start` and `stop` arguments are
+always the first and last elements of the result, not powers applied to some base.
+
+!!! compat "Julia 1.10"
+    This function requires at least Julia 1.10.
+
+# Examples
+```jldoctest
+julia> LogRange(10, 4000, 3)
+3-element LogRange{Float64, Base.TwicePrecision{Float64}}:
+ 10.0, 200.0, 4000.0
+
+julia> ans[2] ≈ sqrt(10 * 4000)  # middle element is the geometric mean
+true
+
+julia> LogRange(√2, 32, 10)
+10-element LogRange{Float64, Base.TwicePrecision{Float64}}:
+ 1.41421, 2.0, 2.82843, 4.0, 5.65685, 8.0, 11.3137, 16.0, 22.6274, 32.0
+
+julia> LogRange(1, 1000, 4) ≈ 10 .^ (0:3)
+true
+
+julia> LogRange(-27, -3f0, 5)  # allows negative numbers
+5-element LogRange{Float32, Float64}:
+ -27.0, -15.5885, -9.0, -5.19615, -3.0
+
+julia> LogRange(1, -1+0im, 5) ≈ cis.(LinRange(0, pi, 5))  # complex numbers
+true
+```
+"""
+struct LogRange{T,X} <: AbstractArray{T,1}
+    start::T
+    stop::T
+    len::Int
+    extra::Tuple{X,X}
+    function LogRange(start::Number, stop::Number, length::Integer)
+        T = float(promote_type(typeof(start), typeof(stop)))
+        a = iszero(start) ? T(NaN) : T(start)
+        b = iszero(stop) ? T(NaN) : T(stop)
+        len = Int(length)
+        if len < 0
+            throw(ArgumentError(LazyString(
+                "LogRange(", start, ", ", stop, ", ", len, "): can't have negative length")))
+        elseif len == 1 && start != stop
+            throw(ArgumentError(LazyString(
+                "LogRange(", start, ", ", stop, ", ", len, "): endpoints differ, while length is 1")))
+        elseif iszero(start) || iszero(stop)
+        elseif T <: Real && (start<0) ⊻ (stop<0)
+            throw(DomainError((start, stop),
+                "LogRange will only return complex results if called with a complex argument"))
+        end
+        ex = if T <: Real && start + stop < 0  # start+stop allows for LogRange(-0.0, -2, 3)
+            _logrange_extra(-a, -b, len)
+        else
+            _logrange_extra(a, b, len)
+        end
+        new{T,typeof(ex[1])}(a, b, len, ex)
+    end
+end
+
+size(r::LogRange) = (r.len,)
+length(r::LogRange) = r.len
+
+first(r::LogRange) = r.start
+last(r::LogRange) = r.stop
+
+function _logrange_extra(a::Number, b::Number, len::Int)
+    loga = log(1.0 * a)  # widen to at least Float64
+    logb = log(1.0 * b)
+    (loga/(len-1), logb/(len-1))
+end
+function _logrange_extra(a::Float64, b::Float64, len::Int)
+    loga = TwicePrecision(Math._log_ext(reinterpret(UInt64, a))...)
```

Comment:
this isn't quite correct. You need to do some extra work for subnormals. I should probably add a `Math._log_ext` for `Float64` that does that for you. The TLDR is that you need to steal the sign bit for an extra bit of exponent to let you renormalize. Also what should happen if `a` and `b` are negative?

Reply:
Ok. Is there an example to look at? Else wrapping it up somewhere would be great. 

> what should happen if a and b are negative?

This part is easy, the constructor has already ensured they are of the same-sign. (Zeros give NaN which is handled elsewhere.)

Reply:
yeah. This is done currently by https://github.com/JuliaLang/julia/blob/bedd14d84e63bc5e1d2be03baf4532e60b88ec9d/base/math.jl#L1125

Reply:
See if this looks right to you. Some very crude tests show high accuracy on subnormals, where previously the result was disastrously wrong:
```julia
function _log_twice64_unchecked(x::Float64)
    xu = reinterpret(UInt64, x)
    if xu < (UInt64(1)<<52) # x is subnormal
        xu = reinterpret(UInt64, x * 0x1p52) # normalize x
        xu &= ~sign_mask(Float64)
        xu -= UInt64(52) << 52 # mess with the exponent
    end
    TwicePrecision(Math._log_ext(xu)...)
end
```
Not sure where such a function ought to live. Surely its tests are in the wrong file right now.

I left off checks for NaN, negative, etc as this is convenient for the present state of LogRange, and the way it produces these:
```julia
julia> LogRange(1,Inf,5)
5-element LogRange{Float64, Base.TwicePrecision{Float64}}:
 1.0, Inf, Inf, Inf, Inf

julia> LogRange(0,3,5)
5-element LogRange{Float64, Base.TwicePrecision{Float64}}:
 NaN, NaN, NaN, NaN, 3.0
```

Reply:
Accuracy improvement:
```julia
julia> rangestats(LogRange, 1e-320, 1e200, 10^5)
(avg = 0.05467, worst = 1)  # now
(avg = 6.26085294268492e15, worst = 9007147337488117)  # before
```
(after fixing `rangestats` not to hit integer overflow from large errors!)

Reply:
`exp(::TwicePrecision)` should return a `TwicePrecision` so I don't love it.

Reply:
Ok. Last commit adds an `_exp_allowing_twice64` for this purpose instead.

Reply:
Do you have further requested changes about log/exp? If not, could you remove the flag?